test/e2e/node_drain_timeout.go

```diff
@@ -0,0 +1,200 @@
+			WaitForMachineDeployments:    input.E2EConfig.GetIntervals(specName, "wait-worker-nodes"),
+		})
+		cluster = applyClusterTemplateResult.Cluster
+		controlplane = applyClusterTemplateResult.ControlPlane
+		machineDeployments = applyClusterTemplateResult.MachineDeployments
+
+		By("Update the nodeDrainTimeout field of the machinedeployment and wait for all machines to be updated")
+
+		nodeDrainTimeoutSecond := 60
+		nodeDrainTimeoutDuration := durationMaker(durationMakerInput{
+			second: float64(nodeDrainTimeoutSecond),
+		})
+		nodeDrainTimeoutMachineDeploymentInterval := convertMachineDeploymentDurationToInterval(nodeDrainTimeoutDuration)
+
+		framework.UpdateNodeDrainTimeoutInMachineDeployment(ctx, framework.UpdateNodeDrainTimeoutInMachineDeploymentInput{
+			ClusterProxy:               input.BootstrapClusterProxy,
+			Cluster:                    cluster,
+			MachineDeployments:         machineDeployments,
+			NodeDrainTimeout:           nodeDrainTimeoutDuration,
+			WaitForMachinesToBeUpdated: input.E2EConfig.GetIntervals(specName, "wait-machine-updated"),
+		})
+
+		By("Add a deployment and podDisruptionBudget to the workload cluster. The deployed pods cannot be evicted in the node draining process.")
+		framework.AddUnevictablePod(ctx, framework.AddUnevictablePodInput{
+			ClusterProxy:                       input.BootstrapClusterProxy,
+			Cluster:                            cluster,
+			MachineDeployments:                 machineDeployments,
+			WaitForDeploymentAvailableInterval: input.E2EConfig.GetIntervals(specName, "wait-deployment-available"),
+		})
+
+		By("Scale the machinedeployment down to zero. If we didn't have the NodeDrainTimeout duration, the node drain process would block this operator.")
+		for _, md := range machineDeployments {
+			framework.ScaleAndWaitMachineDeployment(ctx, framework.ScaleAndWaitMachineDeploymentInput{
+				ClusterProxy:              input.BootstrapClusterProxy,
+				Cluster:                   cluster,
+				MachineDeployment:         md,
+				WaitForMachineDeployments: nodeDrainTimeoutMachineDeploymentInterval,
+				Replicas:                  0,
+			})
+		}
+
+		By("Update nodeDrainTimeout field of the existing and new controlplane machines.")
+		numScaledUpControlPlane := 3
+		framework.UpdateControlplaneNodeDrainTimeout(ctx, framework.UpdateControlplaneNodeDrainTimeoutInput{
+			Controlplane:               controlplane,
+			NodeDrainTimeout:           nodeDrainTimeoutDuration,
+			ClusterProxy:               input.BootstrapClusterProxy,
+			Cluster:                    cluster,
+			ScaleUpTo:                  int32(numScaledUpControlPlane),
+			WaitForMachinesToBeUpdated: input.E2EConfig.GetIntervals(specName, "wait-controlplane-updated"),
+		})
+		By("Deploy workload on the master node. The workload is actually the pods that we deployed above.")
+		framework.DeployWorkloadOnControlplaneNode(ctx, framework.DeployWorkloadOnControlplaneNodeInput{
```

Comment:
Should this be `DeployUnevictablePod` with a Toleration and a NodeSelector?

Reply:
Good suggestion, I will make an update on it.

Reply:
Sorry, I don't get your point here. In `DeployWorkloadOnControlplaneNode` I untainted the master nodes and used NodeSelector to make sure that pods are deployed on all 3 master nodes. Was that what you mean? Or do you mean the description in `By(...)` should be updated?

Reply:
The point here is that DeployUnevictablePod and DeployWorkloadOnControlplaneNode are doing the same job, which is to beploy an unevictable pod; the only difference is that they targets different nodes.
So IMO we should have a single method, DeployUnevictablePod, and add to this method new input parameters (a flag or bare tolerations/selectors) to support deployment on workers and on control-plane nodes

Reply:
I see.  Actually`DeployWorkloadOnControlplaneNode` was needed to untainted the controlplane nodes and then all exiting deployments are redeployed in controlplane nodes. The existing deployments were deployed in previous steps in worker nodes to test the node drain timeout for the `machinedeployment`. After all worker machines have been deleted, these deployments  are waiting until there are available nodes to redeploy them. 
In the newest fix, I have deleted `DeployWorkloadOnControlplaneNode` and moved its functionality into `DeployUnevictablePod`. So now we only one function for deploying unevictable pods.